opensafely/jupyter.py

```diff
@@ -0,0 +1,104 @@
+import argparse
+import json
+import os
+import subprocess
+import sys
+import threading
+import time
+import webbrowser
+from pathlib import Path
+from urllib import request
+
+DESCRIPTION = "Run a jupyter lab notebook using the OpenSAFELY environment"
+
+
+def add_arguments(parser):
+    # we copy a number of standard jupyter lab arguments, and capture them to handle them ourselves
+    parser.add_argument(
+        "--no-browser",
+        default=False,
+        action="store_true",
+        help="Do not attempt to open a browser",
+    )
+    parser.add_argument(
+        "--port",
+        default="8888",
+        help="Port to run on",
+    )
+    # opt into looser argument parsing
+    return True
+
+
+def open_browser(name, port):
+
+    try:
+        # wait for container to be up
+        while True:
+            ps = subprocess.run(
+                ["docker", "inspect", name],
+                capture_output=True,
+                universal_newlines=True,
```

Comment:
Just a by-the-way, but `text=True` is the new way of spelling this as `universal_newlines` was considered not readable enough (kids these day, etc).